### ISSUE TITLE
Shuffle iterator_train by default

### ISSUE THREAD

**Issue by EthanRosenthal** [dnouri/skorch#419]
Please correct me if I am wrong, but it looks like `iterator_train` does not shuffle the dataset by default. This came as a surprise to me when I noticed a degradation in model performance when migrating custom PyTorch training code to `skorch`. I have now used `skorch` on a number of projects and never realized that I should be shuffling`iterator_train`, and I imagine that others have had the same issue.

I was wondering what you think about changing the default behavior of `skorch` to shuffle `iterator_train`? I would be happy to submit a PR for this; or, a warning could be displayed if `iterator_train` does not shuffle.

**Comment by BenjaminBossan** [comment]
Thank you for your proposal. The reason why shuffling is not the default is because [`DataLoader`](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) does not shuffle by default. I agree that often shuffling is the desired behaviour but am not sure if we should change the default. Perhaps a middle ground would be to highlight this behaviour more in the docs and examples, what do you think?

PS: Here is the easiest way to enable shuffling on train data:

https://skorch.readthedocs.io/en/latest/user/FAQ.html#how-do-i-shuffle-my-train-batches

**Comment by EthanRosenthal** [comment]
Fair point - I assumed that was the reasoning behind your decision. I guess the hard thing is competing between PyTorch defaults and sklearn defaults (e.g. [SGDClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html#sklearn.linear_model.SGDClassifier) shuffles by default). Looking through the skorch [tutorials](https://skorch.readthedocs.io/en/latest/user/tutorials.html), it seems that only two tutorials shuffle the data during training, so I wonder how widespread this is.

I understand wanting to keep things as is, though. Highlighting this in the docs would definitely be helpful. Maybe making sure the Quickstart and README both include this parameter with an inline comment explaining why? Also, maybe the examples should be rerun with shuffling (if nothing else, the performance should improve 🙂 )

**Comment by EthanRosenthal** [comment]
Sure thing - I'll try to submit something this week.

**Comment by ZaydH** [comment]
I ran into the same issue @EthanRosenthal mentions.  I had assumed (clearly wrongly) that `skorch` would shuffle by default.  In my mind, many Skorch users may be more nascent in their experience so giving them some (generally) helpful default settings may allow them to come up to be productive faster.  I do get the argument that this is "pytorch territory," and it may be better to keep it consistent.

I agree that the FAQ is a good place to put information like this.  I am not sure I have a better idea.  It may be worth also putting it in more of the `examples` notebooks.  I only see it now in the nuclei notebook.

Just my humble two cents.

**Comment by BenjaminBossan** [comment]
Thanks for the feedback. In addition to that notebook, this is also mentioned in the README and the [docs](https://skorch.readthedocs.io/en/latest/user/quickstart.html). If you find some other place where it could be added, feel free to make a PR.
